base/linalg/dense.jl

```diff
@@ -693,6 +690,309 @@ function inv(A::StridedMatrix{T}) where T
+    end
+    X, Y = exp!(im*A), exp!(-im*A)
+    return (X - Y) / (im * (X + Y))
+end
+
+"""
+    cosh(A::StridedMatrix)
+
+Compute the matrix hyperbolic cosine of a square matrix `A`.
+"""
+function cosh(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(cosh(Hermitian(A)))
+    end
+    return (exp(A) + exp!(-A)) / 2
+end
+
+"""
+    sinh(A::StridedMatrix)
+
+Compute the matrix hyperbolic sine of a square matrix `A`.
+"""
+function sinh(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(sinh(Hermitian(A)))
+    end
+    return (exp(A) - exp!(-A)) / 2
+end
+
+"""
+    tanh(A::StridedMatrix)
+
+Compute the matrix hyperbolic tangent of a square matrix `A`.
+"""
+function tanh(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(tanh(Hermitian(A)))
+    end
+    X, Y = exp(A), exp!(-A)
+    return (X - Y) / (X + Y)
+end
+
+"""
+    acos(A::StridedMatrix)
+
+Compute the inverse matrix cosine of a square matrix `A`.
+
+If `A` is symmetric or Hermitian, its eigendecomposition ([`eigfact`](@ref)) is used to
+compute the inverse cosine. Otherwise, the inverse cosine is determined by using
+[`log`](@ref) and [`sqrt`](@ref).
+
+# Examples
+```jldoctest
+julia> acos(cos(ones(2, 2)))
+2×2 Array{Float64,2}:
+  0.291927  -0.708073
+ -0.708073   0.291927
+```
+"""
+function acos(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(acos(Hermitian(A)))
+    end
+    return -im * log(A + sqrt(A^2 - one(A)))
+end
+
+"""
+    asin(A::StridedMatrix)
+
+Compute the inverse matrix sine of a square matrix `A`.
+
+If `A` is symmetric or Hermitian, its eigendecomposition ([`eigfact`](@ref)) is used to
+compute the inverse sine. Otherwise, the inverse sine is determined by using [`log`](@ref)
+and [`sqrt`](@ref).
+
+# Examples
+```jldoctest
+julia> asin(sin(ones(2, 2)))
+2×2 Array{Float64,2}:
+  0.291927  -0.708073
+ -0.708073   0.291927
+```
+"""
+function asin(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(asin(Hermitian(A)))
+    end
+    return -im * log(im * A + sqrt(one(A) - A^2))
+end
+
+"""
+    atan(A::StridedMatrix)
+
+Compute the inverse matrix tangent of a square matrix `A`.
+
+If `A` is symmetric or Hermitian, its eigendecomposition ([`eigfact`](@ref)) is used to
+compute the inverse tangent. Otherwise, the inverse tangent is determined by using
+[`log`](@ref).
+
+# Examples
+```jldoctest
+julia> atan(tan(ones(2, 2)))
+2×2 Array{Float64,2}:
+  0.291927  -0.708073
+ -0.708073   0.291927
+```
+"""
+function atan(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(atan(Hermitian(A)))
+    end
+    return log((one(A) - im * A) \ (one(A) + im * A)) / 2im
+end
+
+"""
+    acosh(A::StridedMatrix)
+
+Compute the inverse hyperbolic matrix cosine of a square matrix `A`.
+"""
+function acosh(A::StridedMatrix{T}) where T
+    if ishermitian(A)
+        return full(acosh(Hermitian(A)))
+    end
+    return log(A + sqrt(A^2 - one(A)))
```

Comment:
[Aprahamian](http://eprints.ma.man.ac.uk/2432/01/covered/MIMS_ep2016_4.pdf) notes that this is the algorithm used in Octave and comments:

> This function has two weaknesses. First, the formula used for acosh is `acosh A = log(A + sqrt(A² − I))`, which differs from (3.3) (cf. Lemma 3.4) and does not produce the principal branch as we have defined it. Second, the formulas are implemented as calls to logm and sqrtm and so two Schur decompositions are computed rather than one.

The efficiency issue I noted above.  I'm more concerned with the fact that this produces the wrong branch cut.  Are the branch cuts for your other inverse trig functions incorrect as well?

Reply:
The formula they use, which produces the correct branch cut, is `acosh(A) = log(A + sqrt(A-I) * sqrt(A+I))`, which unfortunately requires a second matrix square root.  (This formula makes it even more desirable to compute the Schur factors of A only once.)